File: src/content/docs/topics/cs/coding-problems/math-and-geometry/2013-detect-squares.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class DetectSquares {
    // TODO: implement

    add(point: number[]): void {
        // TODO: implement
    }

    count(point: number[]): number {
        // TODO: implement
        return 0;
    }
}

function _runTests(): void {
    const d = new DetectSquares();
    d.add([3, 10]);
    d.add([11, 2]);
    d.add([3, 2]);
    assert(d.count([11, 10]) === 1);
    assert(d.count([14, 8]) === 0);
    d.add([11, 2]);
    assert(d.count([11, 10]) === 2);

    const d2 = new DetectSquares();
    assert(d2.count([0, 0]) === 0);

    const d3 = new DetectSquares();
    d3.add([0, 0]);
    d3.add([2, 0]);
    d3.add([0, 2]);
    d3.add([2, 2]);
    assert(d3.count([0, 0]) === 1);

    const t0 = performance.now();
    const dBig = new DetectSquares();
    for (let i = 0; i < 1000; i++) {
        dBig.add([i % 32, i % 32]);
        dBig.count([i % 32, (i + 1) % 32]);
    }
    console.log(`perf DetectSquares 1000 add+count ops: ${(performance.now() - t0).toFixed(1)}ms`);

    console.log('all tests pass');
}

_runTests();

File: src/content/docs/topics/cs/coding-problems/math-and-geometry/2013-detect-squares.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

type DetectSquares struct {
	// TODO: implement
}

func Constructor() DetectSquares {
	return DetectSquares{}
}

func (ds *DetectSquares) Add(point []int) {
	// TODO: implement
}

func (ds *DetectSquares) Count(point []int) int {
	// TODO: implement
	return 0
}

func runTests() {
	d := Constructor()
	d.Add([]int{3, 10})
	d.Add([]int{11, 2})
	d.Add([]int{3, 2})
	assert(d.Count([]int{11, 10}) == 1)
	assert(d.Count([]int{14, 8}) == 0)
	d.Add([]int{11, 2})
	assert(d.Count([]int{11, 10}) == 2)
	d2 := Constructor()
	assert(d2.Count([]int{0, 0}) == 0)
	d3 := Constructor()
	d3.Add([]int{0, 0})
	d3.Add([]int{2, 0})
	d3.Add([]int{0, 2})
	d3.Add([]int{2, 2})
	assert(d3.Count([]int{0, 0}) == 1)
	fmt.Println("all tests pass")
}

func main() { runTests() }
